### PR TITLE
switch OLM catalogs to imagestream

### PR DESCRIPTION
**Pull request by sjenning**

### CODE REVIEW

File: cmd/install/assets/hypershift_operator.go [hunk 1]
```diff
@@ -778,6 +778,11 @@ func (o HyperShiftOperatorClusterRole) Build() *rbacv1.ClusterRole {
 				Resources: []string{"*"},
 				Verbs:     []string{"*"},
 			},
+			{
+				APIGroups: []string{"image.openshift.io"},
+				Resources: []string{"*"},
+				Verbs:     []string{"*"},
+			},
```
Comment (csrwng) [inline cmd/install/assets/hypershift_operator.go:785]: Also please add imagestreams to the list of resources to dump:
https://github.com/openshift/hypershift/blob/a50f62208e677dacb234f8c46c3ea64abb0dc4bb/cmd/cluster/core/dump.go#L180-L209

File: control-plane-operator/controllers/hostedcontrolplane/olm/assets/catalog-certified.deployment.yaml [hunk 2]
```diff
@@ -2,6 +2,8 @@ apiVersion: apps/v1
 kind: Deployment
 metadata:
   name: certified-operators-catalog
+  annotations:
+    image.openshift.io/triggers: '[{"from":{"kind":"ImageStreamTag","name":"catalogs:certified-operators"},"fieldPath":"spec.template.spec.containers[?(@.name==\"registry\")].image"}]'
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/olm/assets/catalog-certified.deployment.yaml:6]: We don't pick up annotations from this manifest, only spec:
https://github.com/openshift/hypershift/blob/a50f62208e677dacb234f8c46c3ea64abb0dc4bb/control-plane-operator/controllers/hostedcontrolplane/olm/catalogs.go#L89
Either change the reconcile to also pick up annotations or set this in the reconcile.

Base on [this](https://docs.openshift.com/container-platform/4.11/openshift_images/using-imagestreams-with-kube-resources.html) it looks like you may also need to add

```
alpha.image.policy.openshift.io/resolve-names: '*'
```

**Review by csrwng — COMMENTED** [review]
Looks good, a couple of comments

File: control-plane-operator/controllers/hostedcontrolplane/olm/catalogs.go [hunk 3]
```diff
@@ -86,36 +87,19 @@ func ReconcileRedHatOperatorsDeployment(deployment *appsv1.Deployment, ownerRef
 
 func reconcileCatalogDeployment(deployment *appsv1.Deployment, ownerRef config.OwnerRef, dc config.DeploymentConfig, sourceDeployment *appsv1.Deployment) error {
 	ownerRef.ApplyTo(deployment)
+	deployment.Annotations = sourceDeployment.Annotations
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/olm/catalogs.go:90]: You should preserve other annotations not set in the template. All tests are failing with:

```
WARNING: Object got updated more than one time without a no-op update, this indicates hypershift incorrectly reverting defaulted values
```

```
  &v1.Deployment{
        TypeMeta: {Kind: "Deployment", APIVersion: "apps/v1"},
        ObjectMeta: v1.ObjectMeta{
                ... // 9 identical fields
                DeletionGracePeriodSeconds: nil,
                Labels:                     {"hypershift.openshift.io/managed-by": "control-plane-operator"},
                Annotations: map[string]string{
-                       "deployment.kubernetes.io/revision": "347",
                        "image.openshift.io/triggers":       `[{"from":{"kind":"ImageStreamTag","name":"catalogs:redhat-market`...,
                },
                OwnerReferences: {{APIVersion: "hypershift.openshift.io/v1beta1", Kind: "HostedControlPlane", Name: "example-wq9px", UID: "372fe123-ea6b-47b1-ada8-ad6d874cefcc", ...}},
                Finalizers:      nil,
                ManagedFields:   {{Manager: "hypershift-controlplane-manager", Operation: "Update", APIVersion: "apps/v1", Time: s"2023-01-13 04:43:45 +0000 UTC", ...}, {Manager: "kube-controller-manager", Operation: "Update", APIVersion: "apps/v1", Time: s"2023-01-13 04:43:49 +0000 UTC", ...}},
```

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 4]
```diff
@@ -2572,31 +2579,6 @@ func (r *HostedControlPlaneReconciler) reconcileOperatorLifecycleManager(ctx con
 		}); err != nil {
 			return fmt.Errorf("failed to reconcile catalog rollout rolebinding: %w", err)
 		}
-
-		certifiedOperatorsCronJob := manifests.CertifiedOperatorsCronJob(hcp.Namespace)
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:2576]: Also realized... maybe not as important, but maybe we should have code to delete these jobs if they exist.

Comment (sjenning) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:2576]: I thought about this but we aren't GA yet, and they just cause a little hourly pod churn.  New clusters won't have them.  I thought we could just let them linger for old non-GA clusters.  But if anyone else feels strongly, I'm up for changing my mind.

**Comment by sjenning** [comment]
/hold

still showing no-op failures :thinking:

**Comment by csrwng** [comment]
/hold

Now it's the ImageStream that's hotlooping

```
        Spec: v1.ImageStreamSpec{
                LookupPolicy:          {Local: true},
                DockerImageRepository: "",
                Tags: []v1.TagReference{
                        {
                                ... // 2 identical fields
                                From:            &{Kind: "DockerImage", Name: "registry.redhat.io/redhat/certified-operator-index:v4.12"},
                                Reference:       false,
-                               Generation:      &2,
+                               Generation:      nil,
                                ImportPolicy:    {Scheduled: true},
                                ReferencePolicy: {Type: "Local"},
                        },
                        {
                                ... // 2 identical fields
                                From:            &{Kind: "DockerImage", Name: "registry.redhat.io/redhat/community-operator-index:v4.12"},
                                Reference:       false,
-                               Generation:      &2,
+                               Generation:      nil,
                                ImportPolicy:    {Scheduled: true},
                                ReferencePolicy: {Type: "Local"},
                        },
                        {
                                ... // 2 identical fields
                                From:            &{Kind: "DockerImage", Name: "registry.redhat.io/redhat/redhat-marketplace-index:v4.12"},
                                Reference:       false,
-                               Generation:      &2,
+                               Generation:      nil,
                                ImportPolicy:    {Scheduled: true},
                                ReferencePolicy: {Type: "Local"},
                        },
                        {
                                ... // 2 identical fields
                                From:            &{Kind: "DockerImage", Name: "registry.redhat.io/redhat/redhat-operator-index:v4.12"},
                                Reference:       false,
-                               Generation:      &2,
+                               Generation:      nil,
                                ImportPolicy:    {Scheduled: true},
                                ReferencePolicy: {Type: "Local"},
                        },
                },
        },
```

**Comment by enxebre** [comment]
Nice. Two comments:
[this](https://github.com/openshift/hypershift/pull/2032/files#diff-4edc0a874ce10f9d4f77594c9ff489c3275cb2adc788826a9ea2f48a64c336efR781) will require update rbac in the ci root cluster as soon as this gets merged

Does this PR play well in disconnected envs?

**Pull request merged**